compiler/rustc_lint/src/reexports.rs

```diff
@@ -0,0 +1,76 @@
+use crate::lints::UselessAnonymousReexportDiag;
+use crate::{LateContext, LateLintPass, LintContext};
+use rustc_hir::def::DefKind;
+use rustc_hir::def_id::DefId;
+use rustc_hir::{Item, ItemKind, UseKind};
+use rustc_span::symbol::kw;
+use rustc_span::Span;
+
+declare_lint! {
+    /// The `useless_anonymous_reexport` lint checks if anonymous re-exports
+    /// are re-exports of traits.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(useless_anonymous_reexport)]
+    ///
+    /// mod sub {
+    ///     pub struct Bar;
+    /// }
+    ///
+    /// pub use self::sub::Bar as _;
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Anonymous re-exports are only useful if it's a re-export of a trait
+    /// in case you want to give access to it. If you re-export any other kind,
+    /// you won't be able to use it since its name won't be accessible.
+    pub USELESS_ANONYMOUS_REEXPORT,
+    Warn,
+    "useless anonymous re-export"
+}
+
+declare_lint_pass!(UselessAnonymousReexport => [USELESS_ANONYMOUS_REEXPORT]);
+
+fn emit_err(cx: &LateContext<'_>, span: Span, def_id: DefId) {
+    let article = cx.tcx.def_descr_article(def_id);
+    let desc = cx.tcx.def_descr(def_id);
+    cx.emit_spanned_lint(
+        USELESS_ANONYMOUS_REEXPORT,
+        span,
+        UselessAnonymousReexportDiag { article, desc },
+    );
+}
+
+impl<'tcx> LateLintPass<'tcx> for UselessAnonymousReexport {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'tcx>) {
+        if let ItemKind::Use(path, kind) = item.kind &&
+            !matches!(kind, UseKind::Glob) &&
+            item.ident.name == kw::Underscore &&
+            let Ok(snippet) = cx.tcx.sess.source_map().span_to_snippet(item.span) &&
+            // We only want re-export, not non-reexport "use".
+            (snippet.starts_with("pub ") || snippet.starts_with("pub("))
```

Comment:
I tried a few things but I couldn't find out how to get this information without getting the code snippet directly since the visibility information we can get is only relative and not "this is a pub(crate) item".

Reply:
If I understand correctly, you'd like to only consider use-items as reexports if their visibility is anything other than `pub(self)`, i.e. if their visibility extends to at least the parent module (`pub(super)`, `pub(crate)`, …, `pub`), right? In this case `snippet.starts_with("pub("))` would incorrectly classify `pub(self) use`s as reexports btw.

What about getting the visibility via `tcx.local_visibility(item.owner_id.def_id)` and checking if it is either `Public` or `Restricted(reach)` where `reach != item.owner_id.def_id`? Not sure if that's 100% correct though.

There's `Visibility::is_{accessible_from,at_least}` but I don't they do the thing you need.

Reply:
If there is any form of `pub`, it's considered as a re-export (as far as I know, might be wrong on that). With that in mind, I couldn't use `reach != item.owner_id.def_id` because of `pub(self)`. Welcome to my lost afternoon. :rofl:

Reply:
… but there should be no semantic difference between ` ` (empty/inherited visibility) and `pub(self)` or `pub(in self)`, or am I completely off base here? Haha

Reply:
I was sure there was but you're right, it's not the case. Great catch, it'll greatly simplify this check, thanks!

Reply:
@fmease Just implemented what you suggested and it works like a charm, thanks again. :)